File: src/main/java/com/sedmelluq/discord/lavaplayer/container/adts/AdtsAudioTrack.java
package com.sedmelluq.discord.lavaplayer.container.adts;

import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import com.sedmelluq.discord.lavaplayer.track.BaseAudioTrack;
import com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.InputStream;

/**
 * Audio track that handles an ADTS packet stream
 */
public class AdtsAudioTrack extends BaseAudioTrack {
  private static final Logger log = LoggerFactory.getLogger(AdtsAudioTrack.class);

  private final InputStream inputStream;

  /**
   * @param trackInfo Track info
   * @param inputStream Input stream for the ADTS stream
   */
  public AdtsAudioTrack(AudioTrackInfo trackInfo, InputStream inputStream) {
    super(trackInfo);

    this.inputStream = inputStream;
  }

  @Override
  public void process(LocalAudioTrackExecutor localExecutor) throws Exception {
    AdtsStreamProvider provider = new AdtsStreamProvider(inputStream, localExecutor.getProcessingContext());

    try {
      log.debug("Starting to play ADTS stream {}", getIdentifier());

      localExecutor.executeProcessingLoop(provider::provideFrames, null);
    } finally {
      provider.close();
    }
  }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/container/adts/AdtsAudioTrack.kt
package com.sedmelluq.discord.lavaplayer.container.adts

import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo
import com.sedmelluq.discord.lavaplayer.track.BaseAudioTrack
import com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor
import mu.KotlinLogging
import java.io.InputStream

/**
 * Audio track that handles an ADTS packet stream
 *
 * @param trackInfo   Track info
 * @param inputStream Input stream for the ADTS stream
 */
class AdtsAudioTrack(trackInfo: AudioTrackInfo?, private val inputStream: InputStream) : BaseAudioTrack(trackInfo!!) {
    companion object {
        private val log = KotlinLogging.logger { }
    }

    @Throws(Exception::class)
    override fun process(executor: LocalAudioTrackExecutor) {
        val provider = AdtsStreamProvider(inputStream, executor.processingContext)
        try {
            log.debug { "Starting to play ADTS stream $identifier" }
            executor.executeProcessingLoop({ provider.provideFrames() }, null)
        } finally {
            provider.close()
        }
    }
}
